perceval/backends/kitsune.py

```diff
@@ -0,0 +1,383 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#     Alvaro del Castillo <acs@bitergia.com>
+#
+
+import functools
+import json
+import logging
+import os.path
+
+import requests
+
+from ..backend import Backend, BackendCommand, metadata
+from ..cache import Cache
+from ..errors import CacheError, ParseError
+
+from ..utils import (DEFAULT_DATETIME,
+                     str_to_datetime,
+                     urljoin)
+
+
+logger = logging.getLogger(__name__)
+
+KITSUNE_URL = "https://support.mozilla.org"
+DEFAULT_OFFSET = 0
+
+
+def kitsune_metadata(func):
+    """Kitsune metadata decorator.
+
+    This decorator takes items overrides `metadata` decorator to add extra
+    information related to Kitsune (offset of the question).
+    """
+    @functools.wraps(func)
+    def decorator(self, *args, **kwargs):
+        for item in func(self, *args, **kwargs):
+            item['offset'] = item['data'].pop('offset')
+            yield item
+    return decorator
+
+
+class Kitsune(Backend):
+    """Kitsune backend for Perceval.
+
+    This class retrieves the questions and answers from a
+    Kitsune url. To initialize this class a
+    url could be provided. If not, https://support.mozilla.org will be used.
+
+    Questions and answers are returned from newer to older.
+
+    :param url: Kitsune url
+    :param cache: cache object to store raw data
+    :param origin: identifier of the repository; when `None` or an
+        empty string are given, it will be set to `url` value
+    """
+    version = '0.1.0'
+
+    def __init__(self, url=None, cache=None, origin=None):
+        if not url:
+            url = KITSUNE_URL
+        origin = origin if origin else url
+
+        super().__init__(origin, cache=cache)
+        self.url = url
+        self.client = KitsuneClient(url)
+        self.cache_offset = 0
+
+    @kitsune_metadata
+    @metadata
+    def fetch(self, offset=DEFAULT_OFFSET):
+        """Fetch questions from the Kitsune url.
+
+        :offset: obtain questions after offset
+        :returns: a generator of questions
+        """
+        logger.info("Looking for questions at url '%s' using offset %s",
+                    self.url, str(offset))
+
+        self._purge_cache_queue()
+        # Add to the cache the offset so it can be used to recover from cache
+        self._push_cache_queue(offset)
+        self._flush_cache_queue()
```

Comment:
Removed! I have in mind that the flush should be done when a semantic unit is completed.